File: revenue/hive/commercial-waste-route-operations/web_console.js
"use strict";
const $ = id => document.getElementById(id);
const token = document.querySelector('meta[name="console-key"]').content;
let state = null, busy = false, pending = null, loadedDate = null, selectedInvoice = null;
let stateGeneration = 0, routeGeneration = 0, invoiceGeneration = 0;
const retryKeys = new Map();

function node(tag, text, className) {
  const element = document.createElement(tag);
  if (text !== undefined && text !== null) element.textContent = String(text);
  if (className) element.className = className;
  return element;
}
function say(message, error = false) {
  $("notice").textContent = message;
  $("notice").classList.toggle("error", error);
}
function failure(error) { say(error.message || String(error), true); }
function amount(value, currency = "") { return `${value ?? "unknown"} ${currency} minor units`.replace(/ +/g, " "); }
function button(label, action, secondary = false) {
  const b = node("button", label, secondary ? "secondary" : "");
  b.type = "button";
  b.addEventListener("click", () => Promise.resolve().then(action).catch(failure));
  return b;
}
function empty(target, message) { target.replaceChildren(node("p", message, "empty")); }
function table(target, headers, records) {
  if (!records.length) { empty(target, "No retained records in this view."); return; }
  const t = node("table"), head = node("tr"), thead = node("thead"), body = node("tbody");
  for (const label of headers) head.append(node("th", label));
  thead.append(head); t.append(thead, body);
  for (const record of records) {
    const tr = node("tr");
    for (const value of record) {
      const td = node("td");
      if (value instanceof Node) td.append(value); else td.textContent = String(value ?? "—");
      tr.append(td);
    }
    body.append(tr);
  }
  target.replaceChildren(t);
}
async function request(path, options = {}) {
  const response = await fetch(path, {...options, headers: {"X-Console-Key": token, ...(options.headers || {})}, cache: "no-store"});
  if (!response.ok) {
    let message = `Request failed (${response.status}).`;
    try { const error = await response.json(); message = `${error.error}: ${error.message}`; } catch (_) {}
    throw new Error(message);
  }
  return response;
}
async function json(path, options) { return (await request(path, options)).json(); }
function saveBlob(blob, name) {
  const url = URL.createObjectURL(blob), a = node("a");
  a.href = url; a.download = name; document.body.append(a); a.click(); a.remove();
  setTimeout(() => URL.revokeObjectURL(url), 30000);
}
async function download(params, name) {
  const response = await request(`/api/download?${new URLSearchParams(params)}`);
  saveBlob(await response.blob(), name);
}
function showResult(value) {
  $("result").textContent = JSON.stringify(value, null, 2);
  $("last-result").open = true;
  if (value.op_key) $("operation-key").value = value.op_key;
}
function showPending(value) {
  pending = value;
  $("retry-box").hidden = !value;
  $("retry-key").textContent = value ? value.op_key : "";
}
async function perform(action, args, exactRequest = null) {
  if (busy) return;
  const fingerprint = JSON.stringify({action, args});
  if (!retryKeys.has(fingerprint)) retryKeys.set(fingerprint, `web:${crypto.randomUUID()}`);
  const body = exactRequest || {action, args, op_key: retryKeys.get(fingerprint)};
  if (pending && !exactRequest) {
    throw new Error("Resolve the pending operation first: retry its exact request or look up its key in Activity.");
  }
  busy = true;
  const controls = Array.from(document.querySelectorAll("button"), b => [b, b.disabled]);
  controls.forEach(([b]) => b.disabled = true);
  showPending(body);
  say(`Saving ${action}. Operation key: ${body.op_key}`);
  let saved = false;
  try {
    const result = await json("/api/command", {method: "POST", headers: {"Content-Type": "application/json"}, body: JSON.stringify(body)});
    saved = true;
    showPending(null); showResult(result);
    say(`Saved ${action}. Operation key: ${result.op_key}. Nothing was sent or charged.`);
    if (action === "route") $("route-date").value = body.args.date;
    await refresh();
    if (action === "invoice") await openInvoice(result.result.invoice_id);
    else if ($("route-date").value && (action === "route" || loadedDate)) await loadRoute($("route-date").value);
  } catch (error) {
    if (saved) say(`Change was saved, but refreshing its view failed: ${error.message}. Use Refresh workspace; do not create a replacement operation.`, true);
    else {
      showResult({op_key: body.op_key, error: error.message, request: body});
      say(`${error.message} The original request and key are retained below for exact retry or lookup.`, true);
    }
  } finally {
    busy = false;
    controls.forEach(([b, disabled]) => { if (b.isConnected) b.disabled = disabled; });
    $("print-invoice").disabled = !selectedInvoice;
    $("manifest-fields").disabled = Boolean(state?.initialized);
  }
}
function customerCurrency(id) { return state?.customers.find(c => c.id === id)?.currency || ""; }
function makeInput(label, type = "text", value = "") {
  const wrapper = node("label", label), input = node("input");
  input.type = type; input.value = value; wrapper.append(input);
  return {wrapper, input};
}
function stopCard(stop, serviceDate) {
  const card = node("article", null, "stop");
  card.append(node("span", stop.status, `tag${stop.status === "EXCEPTION_OPEN" ? " alert" : ""}`),
    node("h3", `${stop.customer_name} · ${stop.site_name}`),
    node("p", `${serviceDate} · ${stop.container_label} · ${stop.service_code}`),
    node("p", stop.stop_id, "muted small"),
    node("p", `Scheduled: ${amount(stop.price_minor, customerCurrency(stop.customer_id))}`),
    node("p", `Recorded charge: ${amount(stop.charge_minor, customerCurrency(stop.customer_id))}`));
  if (stop.exception_code) card.append(node("p", `Exception: ${stop.exception_code}`));
  if (stop.resolution) card.append(node("p", `Resolution: ${stop.resolution}${stop.makeup_service_date ? ` · ${stop.makeup_service_date}` : ""}`));
  if (stop.invoice_id) card.append(node("p", `Retained draft: ${stop.invoice_id}`, "muted"));
  if (!state?.business_date || serviceDate > state.business_date) {
    card.append(node("p", "Planning only: this service date has not arrived.", "muted"));
    return card;
  }
  if (stop.status === "PENDING") {
    card.append(button("Record serviced", async () => {
      if (confirm(`Record actual completed service for ${stop.stop_id}?`)) await perform("record", {stop_id: stop.stop_id, outcome: "SERVICED"});
    }));
    const form = node("form"), reason = makeInput("Skipped-service exception code", "text", "BLOCKED_ACCESS");
    reason.input.required = true; reason.input.maxLength = 240;
    const submit = node("button", "Record skipped service", "secondary"); submit.type = "submit";
    form.append(reason.wrapper, submit);
    form.addEventListener("submit", event => { event.preventDefault(); perform("record", {stop_id: stop.stop_id, outcome: "SKIPPED", exception_code: reason.input.value}).catch(failure); });
    card.append(form);
  } else if (stop.status === "EXCEPTION_OPEN") {
    const form = node("form"), label = node("label", "Resolution"), select = node("select");
    for (const [value, title] of [["NO_SERVICE_NO_CHARGE", "No service — no charge"], ["MAKEUP_COMPLETED_BILLABLE", "Actual completed makeup — billable"]]) {
      const option = node("option", title); option.value = value; select.append(option);
    }
    label.append(select);
    const date = makeInput("Actual makeup service date", "date");
    date.input.min = serviceDate; date.input.max = state.business_date; date.wrapper.hidden = true;
    select.addEventListener("change", () => {
      const needed = select.value === "MAKEUP_COMPLETED_BILLABLE";
      date.wrapper.hidden = !needed; date.input.required = needed;
    });
    const submit = node("button", "Resolve retained exception"); submit.type = "submit";
    form.append(label, date.wrapper, submit);
    form.addEventListener("submit", event => {
      event.preventDefault();
      const args = {stop_id: stop.stop_id, resolution: select.value};
      if (select.value === "MAKEUP_COMPLETED_BILLABLE") args.makeup_service_date = date.input.value;
      if (confirm(`Retain this resolution for ${stop.stop_id}? This does not dispatch a vehicle or charge a customer.`)) perform("resolve", args).catch(failure);
    });
    card.append(form);
  }
  return card;
}
async function refresh() {
  const generation = ++stateGeneration;
  const result = await json("/api/state");
  if (generation !== stateGeneration) return;
  state = result;
  $("business-clock").textContent = result.initialized ? `Business date: ${result.business_date} · Retained timezone: ${result.business_timezone}` : "New workspace — import a manifest to begin";
  $("manifest-fields").disabled = result.initialized;
  $("manifest-panel").open = !result.initialized;
  $("setup-state").textContent = result.initialized ? "The retained manifest is initialized. Catalog and timezone cannot be replaced through this console." : "Choose the correct database and initialize it once. The fictional sample is for a separate demonstration database only.";
  for (const id of ["route-date", "period-start", "period-end"]) if (!$(id).value && result.business_date) $(id).value = result.business_date;
  $("period-end").max = result.business_date || "";
  $("period-start").max = result.business_date || "";
  $("metrics").replaceChildren(...[[result.counts.customers, "Customers"], [result.counts.plans, "Recurring plans"], [result.exception_count, "Open exceptions"], [result.counts.invoice_drafts, "Retained drafts"]].map(([n, text]) => {
    const box = node("div", null, "metric"); box.append(node("strong", n), node("span", text)); return box;
  }));
  $("customers").replaceChildren(...result.customers.map(c => { const option = node("option"); option.value = c.id; option.label = `${c.name} · ${c.currency}`; return option; }));
  $("route-history-label").textContent = `Saved route history — showing ${result.routes.length} of ${result.counts.routes} (any other date can be loaded above)`;
  table($("route-history"), ["Date", "Stops", "Pending", "Exceptions", "Open"], result.routes.map(r => [r.service_date, r.stop_count, r.pending, r.exceptions, button("Open route", () => { $("route-date").value = r.service_date; return loadRoute(r.service_date); }, true)]));
  $("exception-count").textContent = `Showing ${result.exceptions.length} of ${result.exception_count} unresolved stops, oldest first. Other stops remain accessible through their route date.`;
  if (result.exceptions.length) $("exception-list").replaceChildren(...result.exceptions.map(s => stopCard(s, s.service_date))); else empty($("exception-list"), "No open exceptions are retained.");
  $("invoice-count").textContent = `Showing ${result.invoices.length} of ${result.counts.invoice_drafts} retained drafts, newest first. Older IDs can be opened below.`;
  table($("invoice-list"), ["Customer", "Period", "Draft total", "Retained ID", "Actions"], result.invoices.map(i => {
    const actions = node("div"); actions.append(button("View", () => openInvoice(i.id), true), button("JSON", () => download({kind: "invoice", id: i.id}, "waste-invoice-DRAFT.json"), true));
    return [i.customer_name, `${i.period_start} – ${i.period_end}`, amount(i.total_minor, i.currency), i.id, actions];
  }));
  const weekdays = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"];
  table($("plan-list"), ["Customer / site", "Container", "Day", "Service", "Price", "Plan ID"], result.plans.map(p => [`${p.customer_name} / ${p.site_name}`, `${p.container_label} (${p.container_type})`, weekdays[p.weekday], p.service_code, amount(p.price_minor, p.currency), p.id]));
  $("event-count").textContent = `Showing ${result.events.length} of ${result.counts.events} events, newest first. Complete history is available as a download.`;
  table($("event-list"), ["Event", "Action", "Entity", "Operation key"], result.events.map(e => [e.id, e.event_type, e.entity_id, button(e.op_key, () => lookupOperation(e.op_key), true)]));
}
async function loadRoute(date) {
  const generation = ++routeGeneration;
  const result = await json(`/api/route?${new URLSearchParams({date})}`);
  if (generation !== routeGeneration || $("route-date").value !== date) return;
  loadedDate = date;
  $("route-summary").textContent = `${result.route_id} · ${result.stop_count} stops · dates and charges are enforced by the existing desk`;
  $("route-downloads").hidden = false;
  $("stops").replaceChildren(...result.stops.map(s => stopCard(s, date)));
}
async function openInvoice(id) {
  const generation = ++invoiceGeneration;
  const result = await json(`/api/invoice?${new URLSearchParams({id})}`);
  if (generation !== invoiceGeneration) return;
  selectedInvoice = id; $("invoice-id").value = id; $("print-invoice").disabled = false;
  const preview = $("invoice-preview"), lines = node("div", null, "table-wrap");
  preview.hidden = false;
  preview.replaceChildren(node("p", "INVOICE DRAFT — NOT SENT / NOT PAID", "draft-heading"), node("h3", result.customer_name), node("p", result.invoice_id, "small"), node("p", `Service period: ${result.period_start} – ${result.period_end}`), node("p", `Retained business date: ${result.business_date}`), node("p", `Draft total: ${amount(result.total_minor, result.currency)}`), lines);
  table(lines, ["Date", "Site / container", "Service / resolution", "Charge (minor units)", "Stop"], result.lines.map(l => [l.service_date, `${l.site_id} / ${l.container_id}`, `${l.service_code} · ${l.resolution || l.status}${l.makeup_service_date ? ` · makeup ${l.makeup_service_date}` : ""}`, l.charge_minor, l.stop_id]));
  preview.append(node("p", `Retained digest: ${result.receipt_digest}`, "small"), node("p", "Prepared from recorded operations. This is not a sent invoice, payment confirmation, or evidence that real-world service occurred.", "muted"), button("Download exact retained draft JSON", () => download({kind: "invoice", id}, "waste-invoice-DRAFT.json"), true));
}
async function lookupOperation(key) {
  const result = await json(`/api/operation?${new URLSearchParams({key})}`);
  showResult(result);
  if (pending?.op_key === key) {
    showPending(null);
    say(result.found ? "The operation is retained. Its saved result is shown in Activity; refreshing the workspace." : "No saved operation exists for that key. The pending request is cleared; correct the inputs or try again.");
    await refresh();
    if (loadedDate && $("route-date").value) await loadRoute($("route-date").value);
  } else say(result.found ? "Saved operation found. See Activity for the exact result." : "No saved operation exists for that key.");
}
function on(id, type, action) { $(id).addEventListener(type, event => { event.preventDefault(); Promise.resolve().then(() => action(event)).catch(failure); }); }
on("refresh", "click", async () => { await refresh(); if (loadedDate) await loadRoute($("route-date").value); say("Workspace refreshed from the retained database."); });
on("route-form", "submit", () => loadRoute($("route-date").value));
on("generate-route", "click", () => { if ($("route-form").reportValidity()) return perform("route", {date: $("route-date").value}); });
$("route-date").addEventListener("change", () => { routeGeneration++; loadedDate = null; $("route-downloads").hidden = true; $("stops").replaceChildren(); $("route-summary").textContent = "Date changed. Load its retained route or generate a plan."; });
for (const b of document.querySelectorAll("[data-route-format]")) b.addEventListener("click", () => { if (loadedDate) download({kind: "route", date: loadedDate, format: b.dataset.routeFormat}, `waste-route-${loadedDate}.${b.dataset.routeFormat === "markdown" ? "md" : b.dataset.routeFormat}`).catch(failure); });
on("manifest-file", "change", async () => {
  const file = $("manifest-file").files[0]; if (!file) return;
  if (file.size > 1024 * 1024) throw new Error("Manifest file exceeds the console's 1 MiB upload limit.");
  $("manifest").value = new TextDecoder("utf-8", {fatal: true}).decode(await file.arrayBuffer());
  say("Original UTF-8 manifest loaded into the editor. It has not been imported.");
});
on("load-sample", "click", async () => { $("manifest").value = await (await request("/api/sample")).text(); say("Fictional sample loaded into the editor only. Use a separate demonstration database."); });
let csvManifestText = null;
on("csv-file", "change", async () => {
  const file = $("csv-file").files && $("csv-file").files[0];
  if (!file) return;
  $("csv-text").value = await file.text();
  say("CSV loaded into the paste box only. Preview before initializing.");
});
on("csv-preview", "click", async () => {
  csvManifestText = null;
  $("csv-download").disabled = true;
  $("csv-preview-out").textContent = "Previewing…";
  const result = await json("/api/csv-preview", {method: "POST", headers: {"Content-Type": "application/json"}, body: JSON.stringify({csv_text: $("csv-text").value, timezone_policy: $("csv-timezone").value})});
  if (typeof result.manifest_text !== "string") throw new Error("Preview did not return manifest text.");
  csvManifestText = result.manifest_text;
  $("manifest").value = result.manifest_text;
  $("csv-download").disabled = false;
  const counts = result.counts || {};
  $("csv-preview-out").textContent = `No structural errors. ${counts.customers} customers, ${counts.sites} sites, ${counts.containers} containers, ${counts.plans} plans. Timezone ${result.business_timezone}. Manifest text is in the editor. Initialize is still a separate step.`;
  say("CSV preview is read-only. Review the manifest, then initialize if this is a new database.");
});
on("csv-download", "click", () => {
  if (!csvManifestText) throw new Error("Preview a valid CSV before downloading its manifest.");
  saveBlob(new Blob([csvManifestText], {type: "application/json"}), "waste-manifest.json");
});
on("manifest-form", "submit", () => { if (confirm("Initialize this database permanently from this manifest and business timezone?")) return perform("import", {manifest_text: $("manifest").value}); });
on("invoice-form", "submit", () => {
  if (confirm("Create an immutable invoice DRAFT and permanently associate its settled stops? Nothing will be sent or charged.")) return perform("invoice", {customer_id: $("invoice-customer").value, period_start: $("period-start").value, period_end: $("period-end").value});
});
on("invoice-lookup", "submit", () => openInvoice($("invoice-id").value));
on("print-invoice", "click", () => { if (selectedInvoice) window.print(); });
on("download-events", "click", () => download({kind: "events"}, "waste-events.json"));
on("operation-lookup", "submit", () => lookupOperation($("operation-key").value));
on("retry", "click", () => { if (pending) return perform(pending.action, pending.args, pending); });
on("save-retry", "click", () => { if (pending) saveBlob(new Blob([JSON.stringify(pending, null, 2)], {type: "application/json"}), "waste-operation-recovery.json"); });
refresh().then(() => say("Workspace opened. Changes require an explicit operator action; invoice output stays draft-only.")).catch(failure);

File: revenue/hive/commercial-waste-route-operations/web_console.py
#!/usr/bin/env python3
"""Loopback browser operator for the existing Commercial Waste Route Desk."""
from __future__ import annotations

import argparse
import csv
import hmac
import io
import json
import secrets
import sqlite3
import sys
from http.server import BaseHTTPRequestHandler, HTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlsplit

from desk_common import DeskError, ValidationError, StateConflict, iso
from desk_migration import WasteRouteDesk

ASSETS = Path(__file__).resolve().parent
MAX_REQUEST = 2 * 1024 * 1024


def strict_json(raw: str):
    def pairs(items):
        value = {}
        for key, item in items:
            if key in value:
                raise ValidationError(f"duplicate JSON member: {key}")
            value[key] = item
        return value

    def bad_constant(value):
        raise ValidationError(f"non-finite JSON number: {value}")

    try:
        return json.loads(raw, object_pairs_hook=pairs, parse_constant=bad_constant)
    except (ValueError, RecursionError) as exc:
        raise ValidationError("invalid or excessively nested JSON") from exc


def browser_value(value):
    """Keep minor-unit integers exact across the browser's Number boundary."""
    if isinstance(value, dict):
        return {
            key: str(item) if key.endswith("_minor") and type(item) is int
            else browser_value(item)
            for key, item in value.items()
        }
    if isinstance(value, list):
        return [browser_value(item) for item in value]
    return value


def rows(conn, sql, params=()):
    return [dict(row) for row in conn.execute(sql, params)]


def workspace(desk):
    """Read-only, internally consistent projection; mutations stay in the engine."""
    conn = desk.conn
    conn.execute("BEGIN")
    try:
        setting = conn.execute(
            "SELECT value FROM workspace_settings WHERE key='business_timezone'"
        ).fetchone()
        counts = {
            table: conn.execute(f"SELECT COUNT(*) FROM {table}").fetchone()[0]
            for table in ("customers", "sites", "containers", "plans", "routes", "invoice_drafts", "events")
        }
        exception_count = conn.execute(
            "SELECT COUNT(*) FROM stops WHERE status='EXCEPTION_OPEN'"
        ).fetchone()[0]
        return {
            "initialized": setting is not None,
            "business_timezone": setting["value"] if setting else None,
            "business_date": desk.business_date().isoformat() if setting else None,
            "counts": counts,
            "customers": rows(conn, "SELECT id,name,currency FROM customers ORDER BY id"),
            "plans": rows(conn, """
                SELECT p.id,p.weekday,p.service_code,p.price_minor,p.active,
                       c.label container_label,c.container_type,s.name site_name,
                       s.customer_id,cu.name customer_name,cu.currency
                FROM plans p JOIN containers c ON c.id=p.container_id
                JOIN sites s ON s.id=c.site_id JOIN customers cu ON cu.id=s.customer_id
                ORDER BY cu.id,s.id,c.id,p.id
            """),
            "routes": rows(conn, """
                SELECT r.service_date,COUNT(st.id) stop_count,
                       SUM(CASE WHEN st.status='PENDING' THEN 1 ELSE 0 END) pending,
                       SUM(CASE WHEN st.status='EXCEPTION_OPEN' THEN 1 ELSE 0 END) exceptions
                FROM routes r LEFT JOIN stops st ON st.route_id=r.id
                GROUP BY r.id ORDER BY r.service_date DESC LIMIT 120
            """),
            "exception_count": exception_count,
            "exceptions": rows(conn, """
                SELECT st.id stop_id,st.status,st.exception_code,st.customer_id,
                       cu.name customer_name,s.name site_name,c.label container_label,
                       p.service_code,p.price_minor,st.charge_minor,r.service_date
                FROM stops st JOIN routes r ON r.id=st.route_id
                JOIN customers cu ON cu.id=st.customer_id JOIN sites s ON s.id=st.site_id
                JOIN containers c ON c.id=st.container_id JOIN plans p ON p.id=st.plan_id
                WHERE st.status='EXCEPTION_OPEN'
                ORDER BY r.service_date,st.sequence,st.id LIMIT 200
            """),
            "invoices": rows(conn, """
                SELECT d.id,d.customer_id,c.name customer_name,d.period_start,
                       d.period_end,d.total_minor,d.currency
                FROM invoice_drafts d JOIN customers c ON c.id=d.customer_id
                ORDER BY d.rowid DESC LIMIT 100
            """),
            "events": rows(conn, """
                SELECT id,op_key,event_type,entity_kind,entity_id
                FROM events ORDER BY id DESC LIMIT 100
            """),
        }
    finally:
        conn.execute("ROLLBACK")


def execute(desk, body):
    if not isinstance(body, dict) or set(body) != {"action", "args", "op_key"}:
        raise ValidationError("request requires exactly action, args, and op_key")
    action, args, op_key = body["action"], body["args"], body["op_key"]
    shapes = {
        "import": ({"manifest_text"}, set()),
        "route": ({"date"}, set()),
        "record": ({"stop_id", "outcome"}, {"exception_code"}),
        "resolve": ({"stop_id", "resolution"}, {"makeup_service_date"}),
        "invoice": ({"customer_id", "period_start", "period_end"}, set()),
    }
    if not isinstance(action, str) or action not in shapes or not isinstance(args, dict):
        raise ValidationError("unknown action or invalid args")
    required, optional = shapes[action]
    if not required.issubset(args) or set(args) - required - optional:
        raise ValidationError(f"invalid fields for {action}")
    for name, value in args.items():
        if not isinstance(value, str) and not (name in optional and value is None):
            raise ValidationError(f"{name} must be text")
    if not isinstance(op_key, str) or not op_key.strip() or len(op_key) > 128:
        raise ValidationError("op_key must be non-empty text of at most 128 characters")
    if action == "import":
        # Parse original source text here, not a Number-rounded browser object.
        return desk.import_manifest(strict_json(args["manifest_text"]), op_key)
    if action == "route":
        return desk.generate_route(args["date"], op_key)
    if action == "record":
        return desk.record_stop(args["stop_id"], args["outcome"], op_key, args.get("exception_code"))
    if action == "resolve":
        return desk.resolve_exception(args["stop_id"], args["resolution"], op_key, args.get("makeup_service_date"))
    return desk.draft_invoice(args["customer_id"], args["period_start"], args["period_end"], op_key)


def spreadsheet_csv(raw):
    """Neutralize formula-like text only in the downloaded CSV, not stored data."""
    output = io.StringIO(newline="")
    writer = csv.writer(output, lineterminator="\n")
    for record in csv.reader(io.StringIO(raw, newline="")):
        writer.writerow([
            "'" + cell if cell.lstrip().startswith(("=", "+", "-", "@")) else cell
            for cell in record
        ])
    return output.getvalue()


class ConsoleServer(HTTPServer):
    def __init__(self, port, desk):
        super().__init__(("127.0.0.1", port), Handler)
        self.desk = desk
        self.token = secrets.token_urlsafe(32)
        self.origin = f"http://127.0.0.1:{self.server_port}"


class Handler(BaseHTTPRequestHandler):
    server_version = "WasteDesk/1"

    def setup(self):
        super().setup()
        self.connection.settimeout(20)

    def log_message(self, *_):
        # Do not put customer identifiers or operation keys in access logs.
        pass

    def local_request(self, api=False):
        if self.headers.get_all("Host", []) != [self.server.origin.removeprefix("http://")]:
            raise PermissionError("Open the printed 127.0.0.1 address directly")
        origin = self.headers.get("Origin")
        if origin is not None and origin != self.server.origin:
            raise PermissionError("Cross-origin requests are not accepted")
        if self.headers.get("Sec-Fetch-Site") not in (None, "none", "same-origin"):
            raise PermissionError("Open the console directly, not inside another site")
        if api and not hmac.compare_digest(self.headers.get("X-Console-Key", ""), self.server.token):
            raise PermissionError("Console session changed; reload this page")

    def send_content(self, content, content_type="application/json; charset=utf-8", status=200, filename=None):
        raw = content.encode("utf-8") if isinstance(content, str) else content
        self.send_response(status)
        self.send_header("Content-Type", content_type)
        self.send_header("Content-Length", str(len(raw)))
        self.send_header("Cache-Control", "no-store")
        self.send_header("X-Content-Type-Options", "nosniff")
        self.send_header("X-Frame-Options", "DENY")
        self.send_header("Referrer-Policy", "no-referrer")
        self.send_header("Content-Security-Policy", "default-src 'self'; script-src 'self'; style-src 'self' 'unsafe-inline'; connect-src 'self'; object-src 'none'; base-uri 'none'; frame-ancestors 'none'; form-action 'none'")
        if filename:
            self.send_header("Content-Disposition", f'attachment; filename="{filename}"')
        self.end_headers()
        self.wfile.write(raw)

    def send_json(self, value, status=200):
        self.send_content(json.dumps(browser_value(value), ensure_ascii=False, allow_nan=False), status=status)

    def query(self, allowed):
        data = parse_qs(urlsplit(self.path).query, keep_blank_values=True, max_num_fields=8)
        if set(data) - set(allowed) or any(len(v) != 1 for v in data.values()):
            raise ValidationError("unexpected or repeated query parameter")
        return {key: value[0] for key, value in data.items()}

    def invoice(self, invoice_id):
        row = self.server.desk.conn.execute(
            "SELECT payload_json FROM invoice_drafts WHERE id=?", (invoice_id,)
        ).fetchone()
        if row is None:
            raise StateConflict("unknown invoice draft")
        return row["payload_json"]

    def get(self):
        path = urlsplit(self.path).path
        self.local_request(api=path.startswith("/api/"))
        desk = self.server.desk
        if path == "/":
            page = (ASSETS / "web_console.html").read_text(encoding="utf-8")
            self.send_content(page.replace("__CONSOLE_TOKEN__", self.server.token), "text/html; charset=utf-8")
        elif path == "/web_console.js":
            self.send_content((ASSETS / "web_console.js").read_bytes(), "text/javascript; charset=utf-8")
        elif path == "/api/sample":
            self.send_content((ASSETS / "sample_manifest.json").read_bytes())
        elif path == "/api/state":
            self.query(())
            self.send_json(workspace(desk))
        elif path == "/api/route":
            date = self.query(("date",)).get("date")
            result = desk.route_snapshot(date)
            claims = dict(desk.conn.execute("""
                SELECT il.stop_id,il.invoice_id FROM invoice_lines il
                JOIN stops st ON st.id=il.stop_id JOIN routes r ON r.id=st.route_id
                WHERE r.service_date=?
            """, (date,)))
            for stop in result["stops"]:
                stop["invoice_id"] = claims.get(stop["stop_id"])
            self.send_json(result)
        elif path == "/api/invoice":
            self.send_json(strict_json(self.invoice(self.query(("id",)).get("id", ""))))
        elif path == "/api/operation":
            key = self.query(("key",)).get("key", "")
            row = desk.conn.execute(
                "SELECT action,result_json FROM operations WHERE op_key=?", (key,)
            ).fetchone()
            self.send_json({"found": row is not None, "op_key": key,
                            "action": row["action"] if row else None,
                            "result": strict_json(row["result_json"]) if row else None})
        elif path == "/api/download":
            q = self.query(("kind", "date", "format", "id"))
            kind = q.get("kind")
            if kind == "route":
                date, fmt = iso(q.get("date"), "date"), q.get("format", "json")
                if fmt == "json":
                    raw, mime, ext = json.dumps(desk.route_snapshot(date), ensure_ascii=False, indent=2), "application/json", "json"
                elif fmt == "csv":
                    raw, mime, ext = spreadsheet_csv(desk.route_csv(date)), "text/csv", "csv"
                elif fmt == "markdown":
                    raw, mime, ext = desk.route_markdown(date), "text/markdown", "md"
                else:
                    raise ValidationError("format must be json, csv, or markdown")
                self.send_content(raw, mime + "; charset=utf-8", filename=f"waste-route-{date}.{ext}")
            elif kind == "invoice":
                self.send_content(self.invoice(q.get("id", "")), filename="waste-invoice-DRAFT.json")
            elif kind == "events":
                self.send_content(json.dumps(desk.event_log(), ensure_ascii=False, indent=2), filename="waste-events.json")
            else:
                raise ValidationError("unknown download kind")
        else:
            self.send_json({"error": "NotFound", "message": "Unknown console route"}, 404)

    def post(self):
        self.local_request(api=True)
        if self.path != "/api/command":
            self.send_json({"error": "NotFound", "message": "Unknown command route"}, 404)
            return
        lengths = self.headers.get_all("Content-Length", [])
        if len(lengths) != 1 or not lengths[0].isdigit() or self.headers.get("Transfer-Encoding"):
            raise ValidationError("one explicit Content-Length is required")
        length = int(lengths[0])
        if not 0 < length <= MAX_REQUEST:
            raise ValidationError("request must be between 1 byte and 2 MiB")
        if self.headers.get_content_type() != "application/json":
            raise ValidationError("Content-Type must be application/json")
        raw = self.rfile.read(length)
        if len(raw) != length:
            raise ValidationError("incomplete request body")
        try:
            body = strict_json(raw.decode("utf-8", errors="strict"))
        except UnicodeError as exc:
            raise ValidationError("request must be valid UTF-8") from exc
        result = execute(self.server.desk, body)
        self.send_json({"ok": True, "op_key": body["op_key"], "result": result})

    def run_request(self, method):
        try:
            method()
        except PermissionError as exc:
            self.send_json({"error": "Forbidden", "message": str(exc)}, 403)
        except ValidationError as exc:
            self.send_json({"error": type(exc).__name__, "message": str(exc)}, 400)
        except DeskError as exc:
            self.send_json({"error": type(exc).__name__, "message": str(exc)}, 409)
        except (BrokenPipeError, ConnectionResetError, TimeoutError):
            pass
        except sqlite3.Error as exc:
            print(f"Waste console database error: {exc}", file=sys.stderr)
            self.send_json({"error": "DatabaseUnavailable", "message": "Database operation failed. Inspect the server log; keep the operation key for an exact retry."}, 503)
        except Exception as exc:
            print(f"Waste console error: {type(exc).__name__}: {exc}", file=sys.stderr)
            self.send_json({"error": "ConsoleError", "message": "Request failed. Keep the operation key and inspect the server log before retrying."}, 500)

    def do_GET(self):
        self.run_request(self.get)

    def do_POST(self):
        self.run_request(self.post)


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--db", required=True, help="Existing or new waste-desk SQLite database")
    parser.add_argument("--port", type=int, default=8786, help="Loopback port (default 8786)")
    args = parser.parse_args()
    if not 0 <= args.port <= 65535:
        parser.error("port must be between 0 and 65535")
    try:
        with WasteRouteDesk(args.db) as desk:
            with ConsoleServer(args.port, desk) as server:
                print(f"Waste Route Desk: {server.origin}/", flush=True)
                print("Local operator only. Invoice drafts do not send or charge anything. Ctrl-C stops the server.", flush=True)
                server.serve_forever()
        return 0
    except KeyboardInterrupt:
        return 0
    except (DeskError, OSError, sqlite3.Error) as exc:
        print(f"Cannot open waste console: {exc}", file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())
